### ISSUE TITLE
How to avoid detecting parts of a target image?

### ISSUE THREAD

**Issue by AliBYetiser** [ultralytics/ultralytics#14230]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am experimenting with yolo-v8x for an object detection, specifically logo detection problem. My training data consists of webpage screenshots and annotations are of brand logos of my interest. I wish to avoid detecting false positive that occur as a result of detecting parts of a logo. For example if I wish to detect the blue lowercase "ultralytics" logo, I get false positives that have images containing similar "ultra" or "lytics" (say with probabilities between 0.5-0.9). Instead, it would be desirable for me if the model did not score parts of any target class not so high, could you please instruct me on which training parameters I need to edit, and is there anything that I should be careful in the training dataset?

I used default training parameters except I've set `fliplr=0` as I believe model would not benefit from seeing images reflected along left right axis.

**Comment by glenn-jocher** [comment]
@AliBYetiser hello,

Thank you for reaching out with your question on avoiding false positives in logo detection using YOLOv8. It's great to see your proactive approach in searching through existing issues and discussions before posting.

To address the issue of detecting parts of a logo as false positives, here are a few strategies you can consider:

1. **Increase Training Data Quality and Quantity**:
   - Ensure that your training dataset includes a diverse set of examples for each logo, including various sizes, orientations, and backgrounds. This helps the model learn the full context of the logos rather than just parts of them.
   - Include negative samples in your dataset, which are images that do not contain any logos. This can help the model learn to distinguish between logos and non-logo content.

2. **Adjust Training Parameters**:
   - **Non-Maximum Suppression (NMS) Threshold**: You can adjust the NMS threshold to reduce the likelihood of detecting overlapping parts of logos. A higher NMS threshold can help suppress detections that are too close to each other.
     ```python
     model.train(data='coco128.yaml', epochs=100, imgsz=640, iou=0.7)
     ```
   - **Confidence Threshold**: Increase the confidence threshold to filter out lower-confidence detections, which are more likely to be false positives.
     ```python
     model.predict(source='path/to/image.jpg', conf=0.5)
     ```

3. **Data Augmentation**:
   - While you have disabled `fliplr`, consider using other augmentation techniques such as rotation, scaling, and color jittering. These augmentations can help the model generalize better and reduce false positives.

4. **Label Smoothing**:
   - Apply label smoothing to soften the target labels, which can help the model become less confident about partial detections.
     ```python
     model.train(data='coco128.yaml', epochs=100, imgsz=640, label_smoothing=0.1)
     ```

5. **Post-Processing**:
   - Implement additional post-processing steps to filter out detections that do not meet certain criteria, such as minimum area or aspect ratio, which can help eliminate partial logo detections.

6. **Model Architecture**:
   - Consider using a more complex model architecture or ensemble methods if the current model is not performing well enough. Sometimes, a single model might not capture all the nuances required for accurate detection.

Here's an example of how you might adjust some of these parameters in your training script:
```python
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8x.yaml')  # build a new model from YAML
model = YOLO('yolov8x.pt')  # load a pretrained model (recommended for training)

# Train the model with adjusted parameters
results = model.train(
    data='coco128.yaml', 
    epochs=100, 
    imgsz=640, 
    iou=0.7, 
    label_smoothing=0.1, 
    conf=0.5
)
```

For more detailed guidance on training settings, you can refer to the [Train Guide](https://docs.ultralytics.com/modes/train).

If you continue to experience issues, please provide a minimum reproducible example of your code and dataset setup. This will help us diagnose the problem more effectively. You can find more information on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any further questions. Happy training! 😊

**Comment by Y-T-G** [comment]
`mosaic=0`

**Comment by glenn-jocher** [comment]
Hello @Y-T-G,

Thank you for your question!

To avoid detecting parts of a target image, setting `mosaic=0` can indeed help by disabling the mosaic data augmentation, which might sometimes lead to partial detections. However, there are additional steps you can take to further refine your model's performance:

1. **Increase Dataset Diversity**: Ensure your dataset includes a wide variety of logo instances in different contexts and backgrounds. This helps the model learn to distinguish the full logo from its parts.

2. **Adjust Non-Maximum Suppression (NMS) and Confidence Thresholds**: Fine-tuning these parameters can help reduce false positives:
   ```python
   model.predict(source='path/to/image.jpg', conf=0.5, iou=0.7)
   ```

3. **Label Smoothing**: This can help the model be less confident about partial detections:
   ```python
   model.train(data='coco128.yaml', epochs=100, imgsz=640, label_smoothing=0.1)
   ```

4. **Post-Processing Filters**: Implement additional filters to discard detections that do not meet certain criteria, such as minimum area or aspect ratio.

If you haven't already, please ensure you are using the latest version of the Ultralytics packages. If the issue persists, providing a minimum reproducible example would be very helpful. You can find more information on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any further questions. Happy training! 😊

**Comment by AliBYetiser** [comment]
Thank you @Y-T-G for suggesting setting `mosaic=0`. As I understand, in data preparation or augmentation if we create images with partial boxes, model learns to detect partial boxes. So if the solution is to avoid that as much as possible, I guess the default on parameters like `translation,` `scale`, `erasing`, and even `auto_augment` (because 'RandAugment' also does scaling) need to be turned off? But that would severely impact performance too?
And I can't find a iou parameter in train settings docs, and would setting it in prediction time really help this problem?

**Comment by Y-T-G** [comment]
It would impact performance, yes. Because object detection is supposed to detect partial objects. `iou` doesn't have anything to do with partial objects.

A better solution is to train a keypoint/pose model. The model will predict whether certain keypoints are visible. If they aren't visible then you can consider it partial.

**Comment by glenn-jocher** [comment]
Hello @Y-T-G-G,

Thank you for your thoughtful insights!

You're correct that disabling augmentations like `translation`, `scale`, `erasing`, and `auto_augment` can impact performance, as these techniques help the model generalize better. However, balancing augmentation and avoiding partial detections is indeed a challenge.

Regarding the `iou` parameter, you're right—it primarily affects the overlap threshold for Non-Maximum Suppression (NMS) during inference and doesn't directly address partial object detection.

A promising alternative, as you mentioned, is to train a keypoint/pose estimation model. This approach can help identify whether key parts of the logo are visible, allowing you to filter out partial detections. YOLOv8 supports pose estimation, and you can leverage this capability to enhance your detection pipeline.

Here's a brief example of how you might set up a pose estimation model:
```python
from ultralytics import YOLO

# Load a pose estimation model
model = YOLO('yolov8-pose.pt')

# Train the model on your dataset
model.train(data='your_dataset.yaml', epochs=100, imgsz=640)
```

For more detailed guidance on pose estimation, you can refer to the [Pose/Keypoint Estimation Documentation](https://docs.ultralytics.com/tasks/pose).

If you encounter any issues or need further assistance, please provide a minimum reproducible example to help us diagnose the problem more effectively. You can find more information on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out with any additional questions. Happy training! 😊
